src/query_router.rs

```diff
@@ -107,16 +110,18 @@ impl QueryRouter {
     }
 
     /// Try to parse a command and execute it.
-    pub fn try_execute_command(&mut self, mut buf: BytesMut) -> Option<(Command, String)> {
-        let code = buf.get_u8() as char;
+    pub fn try_execute_command(&mut self, buf: &BytesMut) -> Option<(Command, String)> {
+        let mut message_cursor = Cursor::new(buf);
+
+        let code = message_cursor.get_u8() as char;
 
         // Only simple protocol supported for commands.
         if code != 'Q' {
             return None;
         }
 
-        let len = buf.get_i32() as usize;
-        let query = String::from_utf8_lossy(&buf[..len - 5]).to_string(); // Ignore the terminating NULL.
+        let _len = message_cursor.get_i32() as usize;
+        let query = message_cursor.read_string().unwrap();
```

Comment:
can we keep `query` as a `&str`? do we need to convert it to an owned `String`

Reply:
I don't recall, but in all likelihood, we can keep it as `&str`.

Reply:
Can't easily return &str, open to suggestions on how to do that though